storage/scylla_api.c

```diff
@@ -0,0 +1,707 @@
+/*
+ * Copyright (C) 2019 BiiLabs Co., Ltd. and Contributors
+ * All Rights Reserved.
+ * This is free software; you can redistribute it and/or modify it under the
+ * terms of the MIT license. A copy of the license can be found in the file
+ * "LICENSE" at the root of this distribution.
+ */
+#include "scylla_api.h"
+#include <assert.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#define SCYLLA_API_LOGGER "SCYLLA_API"
+
+static logger_id_t logger_id;
+
+void scylla_api_logger_init() { logger_id = logger_helper_enable(SCYLLA_API_LOGGER, LOGGER_DEBUG, true); }
+
+int scylla_api_logger_release() {
+  logger_helper_release(logger_id);
+  if (logger_helper_destroy() != RC_OK) {
+    log_critical(logger_id, "%s.\n", SCYLLA_API_LOGGER);
+    return EXIT_FAILURE;
+  }
+  return 0;
+}
+
+typedef struct Select_Where_ {
+  cass_byte_t* bundle;
+  cass_byte_t* address;
+} Select_Where;
+
+typedef enum { WITH_BUNDLE = 0, WITH_BUNDLE_AND_ADDRESS, NUM_OF_SELECT_METHODS } SelectMethod;
+
+typedef struct Select_Query_ {
+  const char* query;
+} Select_Query;
+
+struct scylla_iota_transaction {
+  cass_byte_t bundle[NUM_FLEX_TRITS_BUNDLE];
+  cass_byte_t address[NUM_FLEX_TRITS_ADDRESS];
+  cass_byte_t hash[NUM_FLEX_TRITS_HASH];
+  cass_byte_t message[NUM_FLEX_TRITS_MESSAGE];
+  cass_int64_t value;
+  cass_int64_t timestamp;
+  cass_byte_t trunk[NUM_FLEX_TRITS_TRUNK];
+  cass_byte_t branch[NUM_FLEX_TRITS_BRANCH];
+};
+
+status_t new_scylla_iota_transaction(scylla_iota_transaction_t** obj) {
+  if (NULL == (*obj = (scylla_iota_transaction_t*)malloc(sizeof(scylla_iota_transaction_t)))) {
+    ta_log_error("%s\n", "new transaction fail");
+    return SC_STORAGE_OOM;
+  }
+  return SC_OK;
+}
+
+void free_scylla_iota_transaction(scylla_iota_transaction_t** obj) {
+  free(*obj);
+  return;
+}
+
+status_t set_transaction_hash(scylla_iota_transaction_t* obj, cass_byte_t* hash, size_t length) {
+  if (obj == NULL || hash == NULL || length != NUM_FLEX_TRITS_HASH) {
```

Comment:
For the first two conditions, we mostly use `SC_TA_NULL` or other NULL status code for different modules

Reply:
How about the boundary check?
For example, length != 0 or length != NUM_FLEX_TRITS_HASH
Is there a common MARCO for this?

Reply:
I think we haven't had one. You could define one by yourselves